// respite/lib/handlerServer/pageAction.js
var http = require('http');

var fetch = function(options, cbs) {
	var req = http.request(options, function(res) {
		var statusCode = res.statusCode;
		if (res.statusCode !== 200) {
			cbs && cbs.other(res.statusCode, res.headers);
		} else {
			var body = "";
			res.on('data', function(chunk) {
				body += chunk;
			});
			res.on('end', function(chunk) {
				cbs && cbs.success(body, res.headers, statusCode);
			});
		}
	});

	req.on('error', function(e) {
		console.log('problem with request: ' + e.message);
		cbs && cbs.fail(e);
	});

	req.end();
}

module.exports = function(req, res, cbs) {
	console.log("[info pageAction] " + req.url);
	var hostName = req.headers["host"].split(":")[0];
	var port = req.headers["host"].split(":")[1] || 80;
	fetch({
		hostname: hostName,
		port: port,
		path: req.url,
		method: req.method,
		headers: req.headers
	}, {
		success: function(body, resHeaders, statusCode) {
			res.writeHead(statusCode, resHeaders);
			if (cbs && cbs.handleBody) {
				body = cbs.handleBody(body, resHeaders);
			}
			res.end(body);
		},
		fail: function(e) {
			res.end(e);
		},
		other: function(statusCode, headers) {
			res.writeHead(statusCode, headers);
			res.end();
		}
	});
}
